core/merkleTree.go

```diff
@@ -36,14 +36,14 @@ type Tree struct {
 // NewTree creates a merkle tree from given leaves
 // the order of leaves will be preserved
 func NewTree(leaves []ByteArray) (newTree Tree, err error) {
+	if len(leaves)%2 != 0 {
+		leaves = append(leaves, ZeroLeaf)
+	}
```

Comment:
Here we handled the odd number of leaves only in the bottom level. But the odd number of nodes should be handled at every level.

For example, say we have 9 leaves to merklise. This makes a 4 level tree (not counting the root).
1. We add defaultHash[0] in level 0 leaves. After the ascend, level 1 has 5 nodes
2. We add defaultHash[1] in level 1 nodes to make them 6 nodes. After the ascend, level 2 has 3 nodes.
3. We add defaultHash[2] in level 2 nodes to make them 4 nodes. After the ascend, level 3 has 2 nodes.

I suggest this logic is better handled in the `ascend` method.